### PR TITLE
[SPARK-27268][SQL] Add map_keys and map_values support in nested schema pruning.

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

We need to add `map_keys` and `map_values` into `ProjectionOverSchema` to support those methods in nested schema pruning. This also adds end-to-end tests to SchemaPruningSuite.

## How was this patch tested?

Added tests.

### CODE REVIEW

**Comment by viirya** [comment]
cc @dongjoon-hyun @maropu @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala [hunk 1]
```diff
@@ -253,6 +257,20 @@ abstract class SchemaPruningSuite
...
+    val query = sql("select map_keys(relations).middle[0], p from contacts")
+    checkScan(query,
+    "struct<relations:map<struct<middle:string>,string>>")
+    checkAnswer(query, Row("Y.", 1) :: Row("X.", 1) :: Row(null, 2) :: Row(null, 2) :: Nil)
+  }
+
+  testSchemaPruning("select nested field from a complex map value using map_values") {
+    val query = sql("select map_values(relatives).middle[0], p from contacts")
+    checkScan(query,
+      "struct<relatives:map<string,struct<middle:string>>>")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala]: `checkScan(query, "struct<relatives:map<string,struct<middle:string>>>")`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala [hunk 2]
```diff
@@ -253,6 +257,20 @@ abstract class SchemaPruningSuite
     checkAnswer(query, Row(1) :: Nil)
   }
 
+  testSchemaPruning("select nested field from a complex map key using map_keys") {
+    val query = sql("select map_keys(relations).middle[0], p from contacts")
+    checkScan(query,
+    "struct<relations:map<struct<middle:string>,string>>")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala]: nit: `    checkScan(query, "struct<relations:map<struct<middle:string>,string>>")`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ProjectionOverSchema.scala [hunk 3]
```diff
@@ -49,6 +49,10 @@ case class ProjectionOverSchema(schema: StructType) {
               s"unmatched child schema for GetArrayStructFields: ${projSchema.toString}"
             )
         }
+      case MapKeys(child) =>
+        getProjection(child).map { projection => MapKeys(projection) }
+      case MapValues(child) =>
+        getProjection(child).map { projection => MapValues(projection) }
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ProjectionOverSchema.scala]: All the extraction expressions listed in `SelectedField` basically should appear in this file?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ProjectionOverSchema.scala]: The expressions which can be part of an extractor chain, should appear in `ProjectionOverSchema`.

**Comment by dongjoon-hyun** [comment]
cc @dbtsai

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master. Thank you, @viirya , @maropu , @dbtsai .

**Pull request closed without merging**